### ISSUE TITLE
Is there any way to return iterator to feed to fastapi StreamingResponse?

### ISSUE THREAD

**Issue by Satyapriya707** [KoljaB/RealtimeTTS#39]
Hi,

I want to know if there's any way to get an iterator over chunks so that I'll be able to do like this -

chunks = stream.<some_way_to_get_chunk_iterator>

def gen():
    for chunk in chunks:
        yield chunk
        
then in fastapi return - StreamingResponse(gen(), media_type="audio/wav")

Thanks

**Comment by TerryWong1024** [comment]
@Satyapriya707 Do you resolve it? i encountered the same problem, please specify if you resolved it.

**Comment by KoljaB** [comment]
Some - very raw - demo code to get you started (I know I should publish a full fastapi realtime TTS server):

```python
from RealtimeTTS import TextToAudioStream, CoquiEngine
from fastapi import FastAPI
from fastapi.responses import StreamingResponse
from queue import Queue
import threading
import uvicorn

app = FastAPI()
audio_queue = Queue()

def on_audio_chunk(chunk):
    audio_queue.put(chunk)

def play_text_to_speech(stream, text):
    stream.feed(text)
    stream.play(on_audio_chunk=on_audio_chunk, muted=True)
    audio_queue.put(None)

def audio_chunk_generator():
    while True:
        chunk = audio_queue.get()
        if chunk is None:
            break
        yield chunk

@app.get("/")
def read_root():
    text_to_speak = "This can be played."
    threading.Thread(target=play_text_to_speech, args=(stream, text_to_speak), daemon=True).start()
    return StreamingResponse(audio_chunk_generator(), media_type="audio/wav")

if __name__ == '__main__':
    print("Initializing TTS Engine...")
    engine = CoquiEngine()
    stream = TextToAudioStream(engine)
    
    print("Starting server ...")
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

Testclient:
```
import pyaudio
import requests

pyaudio_instance = pyaudio.PyAudio()
stream = pyaudio_instance.open(format=pyaudio.paInt16, channels=1, rate=24000, output=True)

def get_tts_audio():
    print("Requesting audio ...")
    url = "http://localhost:8000/"
    response = requests.get(url, stream=True)

    if response.status_code == 200:
        stream.start_stream()

        for chunk in response.iter_content(chunk_size=8192):
            if chunk:
                stream.write(chunk)

        stream.stop_stream()
        stream.close()
    else:
        print("Error:", response.status_code)

if __name__ == "__main__":
    get_tts_audio()
```

**Comment by TaisukeNigo** [comment]
How to write if using H5 page as the client?

**Comment by TerryWong1024** [comment]
> Some - very raw - demo code to get you started (I know I should publish a full fastapi realtime TTS server):
> 
> ```python
> from RealtimeTTS import TextToAudioStream, CoquiEngine
> from fastapi import FastAPI
> from fastapi.responses import StreamingResponse
> from queue import Queue
> import threading
> import uvicorn
> 
> app = FastAPI()
> audio_queue = Queue()
> 
> def on_audio_chunk(chunk):
>     audio_queue.put(chunk)
> 
> def play_text_to_speech(stream, text):
>     stream.feed(text)
>     stream.play(on_audio_chunk=on_audio_chunk, muted=True)
>     audio_queue.put(None)
> 
> def audio_chunk_generator():
>     while True:
>         chunk = audio_queue.get()
>         if chunk is None:
>             break
>         yield chunk
> 
> @app.get("/")
> def read_root():
>     text_to_speak = "This can be played."
>     threading.Thread(target=play_text_to_speech, args=(stream, text_to_speak), daemon=True).start()
>     return StreamingResponse(audio_chunk_generator(), media_type="audio/wav")
> 
> if __name__ == '__main__':
>     print("Initializing TTS Engine...")
>     engine = CoquiEngine()
>     stream = TextToAudioStream(engine)
>     
>     print("Starting server ...")
>     uvicorn.run(app, host="0.0.0.0", port=8000)
> ```
> 
> Testclient:
> 
> ```
> import pyaudio
> import requests
> 
> pyaudio_instance = pyaudio.PyAudio()
> stream = pyaudio_instance.open(format=pyaudio.paInt16, channels=1, rate=24000, output=True)
> 
> def get_tts_audio():
>     print("Requesting audio ...")
>     url = "http://localhost:8000/"
>     response = requests.get(url, stream=True)
> 
>     if response.status_code == 200:
>         stream.start_stream()
> 
>         for chunk in response.iter_content(chunk_size=8192):
>             if chunk:
>                 stream.write(chunk)
> 
>         stream.stop_stream()
>         stream.close()
>     else:
>         print("Error:", response.status_code)
> 
> if __name__ == "__main__":
>     get_tts_audio()
> ```

@KoljaB WE Can't Wait, I firmly believe that a vast number of individuals are eagerly waiting for this enhancement, and it is indeed a genuine, urgent requirement for the majority of real-world scenarios.

**Comment by KoljaB** [comment]
I can imagine lots of people want that. I am working on this with a high priority atm. Will probably be ready in few days. Pls remember tho guys I am doing all this for free so you can build products on top. I understand how urgent need for this may be, pls have a bit patience tho so things can be done right.

**Comment by criminact** [comment]
Does the FastAPI server handles concurrency? If made 2 requests at the same time, it gives all the chunks to the 2nd request.

**Comment by KoljaB** [comment]
No, unfortunately not. Some of the TTS Engines don't support that.

**Comment by criminact** [comment]
Okay. This is my code for handling concurrency on the server side using OpenAI's Engine, This should be okay right? You think anything that can be changed here for the better? - 

Use-case:
I want to stream chunks from OpenAIEngine

Code Chunk:
`
    ...

class TTSRequestHandler:
    def __init__(self, engine):
        self.engine = engine
        self.audio_queue = Queue()
        self.stream = TextToAudioStream(engine)
        self.speaking = False

    def on_audio_chunk(self, chunk):
        self.audio_queue.put(chunk)

    def generator(self, query):
        completion = groq_client.chat.completions.create(
            model="llama3-8b-8192",
            messages=[
                {"role": "user", "content": query}
            ],
            max_tokens=1024,
            stream=True
        )

        for chunk in completion:
            token = chunk.choices[0].delta.content
            if token:
                yield token

    def play_text_to_speech(self, text):
        self.speaking = True
        try:
            self.stream.feed(self.generator(text))
            logging.debug(f"Playing audio for text: {text}")
            print(f"Synthesizing: \"{text}\"")
            self.stream.play(on_audio_chunk=self.on_audio_chunk, muted=True)
            self.audio_queue.put(None)
        finally:
            self.speaking = False

    def audio_chunk_generator(self, send_wave_headers):
        first_chunk = False
        try:
            while True:
                chunk = self.audio_queue.get()
                if chunk is None:
                    print("Terminating stream")
                    break
                if not first_chunk:
                    if send_wave_headers:
                        print("Sending wave header")
                        yield create_wave_header_for_engine(self.engine)
                    first_chunk = True
                print("Sending chunk")
                yield chunk
        except Exception as e:
            print(f"Error during streaming: {str(e)}")

    ...
    request_handler = TTSRequestHandler(OpenAIEngine())
    browser_request = True

    if play_text_to_speech_semaphore.acquire(blocking=False):
        try:
            if not request_handler.speaking:
                threading.Thread(
                    target=request_handler.play_text_to_speech,
                    args=(query,),
                    daemon=True).start()
        finally:
            play_text_to_speech_semaphore.release()

    print(request_handler.speaking)

        return StreamingResponse(
            request_handler.audio_chunk_generator(browser_request),
            media_type="audio/wav"
        )`

**Comment by KoljaB** [comment]
Looks okay at first glance. 

```python
            self.stream.play(on_audio_chunk=self.on_audio_chunk, muted=True)
            self.audio_queue.put(None)
```
I think I would change this code to use play_async. I'd use the on_audio_stream_stop callback to write audio_queue.put(None). Then the main python thread would not block and you could still react to user input. 

Other code looks fine for me.

**Comment by criminact** [comment]
Thank you so much. Do you want me to put this example up in the repo? (In case people are looking for concurrent request handling?)

**Comment by KoljaB** [comment]
Yes, please. I think there's value for others.

**Comment by criminact** [comment]
I will put it up as soon as I have some bandwidth

**Comment by criminact** [comment]
@KoljaB, Please review - 
https://github.com/KoljaB/RealtimeTTS/pull/109

Note: I have commented out coqui engine in the Supported Engines list for this server (it doesn't support multiple requests at once right?)
